torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -171,6 +171,34 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
   }
 }
 
+void eliminatePackedSequence(std::shared_ptr<Graph>& graph) {
+  for (auto it = graph->begin(); it != graph->end(); ++it) {
+    auto* n = *it;
+
+    // for any PadPacked, if we can trace the PackedSequence back to
```

Comment:
I'm not sure this transformation is sound. Can you give an example trace where it's necessary to walk across inputs to successfully eliminate the packing?

Reply:
well, the whole point of the packed sequence is that there's an RNN in the middle. So it will be (Pack -> RNN -> Unpack). I'll add an example.

Reply:
Here's an example export of 3-layer Elman RNN, using packed sequences. (try your best to ignore all the concats and slices cluttering it up, they aren't relevant here)

WIth peephole optimizations turned off, we get

```
graph(%0 : Float(5, 2, 7)
      %1 : Float(3, 2, 13)
      %2 : Long(2)
      %3 : Float(13, 7)
      %4 : Float(13, 13)
      %5 : Float(13)
      %6 : Float(13)
      %7 : Float(13, 13)
      %8 : Float(13, 13)
      %9 : Float(13)
      %10 : Float(13)
      %11 : Float(13, 13)
      %12 : Float(13, 13)
      %13 : Float(13)
      %14 : Float(13)) {
  %15 : Float(10, 7), %16 : Long(5) = PackPadded(%0, %2), scope: RnnModelWithPackedSequence
  %17 : UNKNOWN_TYPE = Constant[value= 10  10  10  10  10  10  10 [ CPUIntTensor{7} ]](), scope: RnnModelWithPackedSequence/RNN[model]
  %18 : UNKNOWN_TYPE = Concat[axis=0](%5, %6), scope: RnnModelWithPackedSequence/RNN[model]
  %19 : UNKNOWN_TYPE = Slice[axes=[0], ends=[1], starts=[0]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %20 : UNKNOWN_TYPE, %21 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%15, %3, %4, %18, %17, %19), scope: RnnModelWithPackedSequence/RNN[model]
  %22 : UNKNOWN_TYPE = Concat[axis=0](%9, %10), scope: RnnModelWithPackedSequence/RNN[model]
  %23 : UNKNOWN_TYPE = Slice[axes=[0], ends=[2], starts=[1]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %24 : UNKNOWN_TYPE, %25 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%20, %7, %8, %22, %17, %23), scope: RnnModelWithPackedSequence/RNN[model]
  %26 : UNKNOWN_TYPE = Concat[axis=0](%13, %14), scope: RnnModelWithPackedSequence/RNN[model]
  %27 : UNKNOWN_TYPE = Slice[axes=[0], ends=[3], starts=[2]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %28 : Float(10, 13), %29 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%24, %11, %12, %26, %17, %27), scope: RnnModelWithPackedSequence/RNN[model]
  %30 : Float(5, 2, 13), %31 : Long(2) = PadPacked(%28, %16), scope: RnnModelWithPackedSequence
  return (%30);
}
```

with peephole optimizations turned back on, we get

```
graph(%0 : Float(5, 2, 7)
      %1 : Float(3, 2, 13)
      %2 : Long(2)
      %3 : Float(13, 7)
      %4 : Float(13, 13)
      %5 : Float(13)
      %6 : Float(13)
      %7 : Float(13, 13)
      %8 : Float(13, 13)
      %9 : Float(13)
      %10 : Float(13)
      %11 : Float(13, 13)
      %12 : Float(13, 13)
      %13 : Float(13)
      %14 : Float(13)) {
  %15 : UNKNOWN_TYPE = Constant[value= 10  10  10  10  10  10  10 [ CPUIntTensor{7} ]](), scope: RnnModelWithPackedSequence/RNN[model]
  %16 : UNKNOWN_TYPE = Concat[axis=0](%5, %6), scope: RnnModelWithPackedSequence/RNN[model]
  %17 : UNKNOWN_TYPE = Slice[axes=[0], ends=[1], starts=[0]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %18 : UNKNOWN_TYPE, %19 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%0, %3, %4, %16, %15, %17), scope: RnnModelWithPackedSequence/RNN[model]
  %20 : UNKNOWN_TYPE = Concat[axis=0](%9, %10), scope: RnnModelWithPackedSequence/RNN[model]
  %21 : UNKNOWN_TYPE = Slice[axes=[0], ends=[2], starts=[1]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %22 : UNKNOWN_TYPE, %23 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%18, %7, %8, %20, %15, %21), scope: RnnModelWithPackedSequence/RNN[model]
  %24 : UNKNOWN_TYPE = Concat[axis=0](%13, %14), scope: RnnModelWithPackedSequence/RNN[model]
  %25 : UNKNOWN_TYPE = Slice[axes=[0], ends=[3], starts=[2]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %26 : Float(10, 13), %27 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%22, %11, %12, %24, %15, %25), scope: RnnModelWithPackedSequence/RNN[model]
  return (%26);
}
```

Reply:
So, what it looks like to me, is that your optimization is validated by the following equations (in a simplified setting where these ops all have one argument/output; in reality the side conditions are slightly more complicated):

```
PadPacked(RNN(x)) == RNN(PackPadded(x))
PadPacked(PackPadded(x)) = Id
```

So it seems more sound to me to structure the optimization pass this way:

1. Traverse the graph looking for occurrences of `PadPacked`
2. If the output of `PadPacked` is used by the first argument of `RNN`, "push" it past the RNN, turning into a `PackPadded` on the first output of the RNN
3. If the use of a `PadPacked` is a `PackPadded`, eliminate the pair of operations.

There's some leeway in whether or not you should require these to be single-use or not before the optimization fires, in reality they are probably all single use.

How does this sound?

Reply:
That sounds pretty good, I'll try out that approach